### PR TITLE
Refactor the repo

### PR DESCRIPTION
**Pull request by nwu63**
## Purpose
This is an attempt at addressing #50. I have taken option 2 here, i.e. elevate classes and functions in each sub-module via its `__init__` file. However, for backwards compatibility, I have kept the root `__init__` file updated. This means that most imports should not break, except for the following instances:
- `adflow/tests/reg_tests/convert_old_refs.py:from baseclasses.BaseRegTest import convertRegFileToJSONRegFile`
- `adflow/tests/reg_tests/reg_test_utils.py:from baseclasses.BaseRegTest import getTol`

In the first case, the file `convert_old_refs.py` can probably be just removed, since it's no longer needed. The second one cannot be fixed easily I think, so we will have to patch ADflow (preferably the same day this is merged).

All the rest of the `baseclasses` imports should be okay. However, if we merge this, my intention is to slowly fix all the repos to reference the submodules when importing, e.g. `from baseclasses.problems import AeroProblem`. Then after that's all done, we should be able to get rid of the imports in the root `__init__` file.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
All existing tests pass.

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Review by sseraj — COMMENTED** [review]
Looks good to me. Just one comment

File: baseclasses/__init__.py [hunk 1]
```diff
@@ -1,27 +1,24 @@
...
+    EngineProblem,
+    FieldPerformanceProblem,
+    LGProblem,
+)
+
+from .solvers import BaseSolver, AeroSolver
+
+from .utils import getPy3SafeString
+
+from .testing import BaseRegTest, require_mpi
```
[the remark this answers is not in the document]
Comment (nwu63) [reply baseclasses/__init__.py:24]: Good point, I had thought that some external repos were importing it but that is not the case. I will change this back.

On a related note, I would prefer if we slowly got rid of this init file, and instead relied on the init files in the sub directories. E.g. `from baseclasses.problems import AeroProblem`. But I see no way to throw deprecation warnings or anything like that to ask people to update.

**Review by marcomangano — DISMISSED** [review]
Looks good, I guess we need a nightly build to be 100% sure that there are no other spurious imports, but it should be a very quick fix anyway.

I am onboard with the refactoring of the other repos to make the main `__init__` not relevant anymore, although I do not have such strong feelings and I think the "dual approach" of this PR is pretty neat. As you commented, I think the real hassle will be to change the runscripts, but that will be an issue for the individual users. I would not worry too much about that... it's a very quick update 

P.S. How bad would it be to change the second import to: `adflow/tests/reg_tests/reg_test_utils.py:from baseclasses import BaseRegTest.getTol as getTol` :grin:

**Comment by nwu63** [comment]
> Looks good, I guess we need a nightly build to be 100% sure that there are no other spurious imports, but it should be a very quick fix anyway.
> 
> I am onboard with the refactoring of the other repos to make the main `__init__` not relevant anymore, although I do not have such strong feelings and I think the "dual approach" of this PR is pretty neat. As you commented, I think the real hassle will be to change the runscripts, but that will be an issue for the individual users. I would not worry too much about that... it's a very quick update
> 
> P.S. How bad would it be to change the second import to: `adflow/tests/reg_tests/reg_test_utils.py:from baseclasses import BaseRegTest.getTol as getTol` grin

Yeah I grepped our repos but I can't be 100% sure that this covers everything. I'm okay with not removing the init since it does not cause any conflicts.

As for `getTol`, I'm not 100% sure that works, I was thinking of `from baseclasses.testing.BaseRegTest import getTol`. But this does highlight the problems you encounter when the module and class have the same name, so it's never clear what `BaseRegTest` is referring to. Is this something we want to fix here by renaming modules which clash with the class name? This should not be backwards incompatible since most of the time users directly import the class anyways.

**Comment by marcomangano** [comment]
> But this does highlight the problems you encounter when the module and class have the same name, so it's never clear what `BaseRegTest` is referring to.

What are you referring specifically to here? I am onboard with some additional renaming within this PR if nothing breaks.

**Comment by nwu63** [comment]
> What are you referring specifically to here? I am onboard with some additional renaming within this PR if nothing breaks.

For example, `from baseclasses.testing import BaseRegTest`, is that importing the class or the module? Because we have the `__init__` it's importing the module, but it's not very obvious since it's like overwriting a variable, it's shadowing something else.

If we renamed the file (e.g. module) that could avoid this issue. I believe this is why we have things like `pyAeroProblem` as the filename when the class is just called `AeroProblem`.

**Comment by nwu63** [comment]
I have now done the following two updates:
- renamed `BaseRegTest.py` to `pyRegTest.py` to avoid confusion. This should not cause any backwards-incompatible changes
- promoted `getTol` so now either `from baseclasses import getTol` or `from baseclasses.testing import getTol` will work.

**Review by sseraj — APPROVED** [review]

**Review by marcomangano — APPROVED** [review]
Sorry for the late review, very good stuff, let's move ahead (and keep monitoring the buildbot for possible "hotfixes")

**Pull request merged**